generator/parser/idl_interface.ts

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright (c) 2017 The Bacardi Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import IDLDefinition from './idl_definition';
+import IDLIdentifier from './idl_identifier';
+
+// FIXME(zino): We should consider attribute and operation concept.
+
+class Argument implements IDLIdentifier {
+  readonly type: string;
+  readonly name: string;
+
+  constructor(raw_argument_info: {}) {
+    this.type = raw_argument_info['idlType']['idlType'];
+    this.name = raw_argument_info['name'];
+  }
+}
+
+class InterfaceConstructor implements IDLIdentifier {
+  readonly type: string;
+  readonly name: string;
+  readonly arguments: Argument[];
+
+  constructor(raw_constructor_info: {}) {
+    this.type = raw_constructor_info['name'];
+    this.name = raw_constructor_info['name'];
+    this.arguments = [];
+    raw_constructor_info['arguments'].forEach(argument => {
+      this.arguments.push(new Argument(argument));
+    });
+  }
+}
+
+class InterfaceMember implements IDLIdentifier {
+  readonly type: string;
+  readonly name: string;
+  readonly arguments?: Argument[];
+  readonly is_static: boolean;
+
+  constructor(raw_member_info: {}) {
+    this.type = raw_member_info['idlType']['idlType'];
+    this.name = raw_member_info['name'];
+    this.is_static = raw_member_info['static'];
+    if (raw_member_info['type'] == 'operation') {
+      this.arguments = [];
+      raw_member_info['arguments'].forEach(argument => {
+        this.arguments.push(new Argument(argument));
+      });
+    } else {
+      this.arguments = null;
+    }
+  }
+}
+
+export default class IDLInterface extends IDLDefinition {
+  constructors: InterfaceConstructor[];
+  members: InterfaceMember[];
+
+  constructor(raw_idl_interface_info: {}) {
+    super(raw_idl_interface_info['name'], raw_idl_interface_info);
+
+    this.members = [];
+    raw_idl_interface_info['members'].forEach(raw_member_info => {
+      this.members.push(new InterfaceMember(raw_member_info));
+    });
+
+    this.constructors = [];
+    raw_idl_interface_info['extAttrs'].forEach(raw_constructor_info => {
```

Comment:
I think we should decide whether to use ```extAttrs``` or ```constructors```.

Reply:
Sure, my thought is that..

According to the WebIDL spec, the constructor is defined in extended attributes.
Nevertheless, I thought the constructor concept is near by special operation(function).
So, I defined it as OperationMember but sometimes it can be different with normal ```operation(function)```. So, eventually, I defined it as InterfaceConstructor.

The WebIDL2 parser (we used internally) provides the constructor information as extAttrs and it conceptually follows the spec well. But it might make our implementation difficult. For example, there are two extended attributes as follows:
- Constructor
- ImplementedAs: This is generally used to specify explicit function name in cpp side mapping with IDL function name.

In the spec, both of them belong to extended attributes but their contents are each very different. In other words, their implementations are very different. So, I think it's better to distinguish them. That's why the parser layer exists. If not so, we don't need the layer and we would just use parsed data from WebIDL2 parser directly.

Reply:
Okay I understand your means. It's good idea.